Mathlib/RingTheory/AlgebraicIndependent/Transcendental.lean

```diff
@@ -78,42 +83,53 @@ theorem AlgebraicIndependent.option_iff (hx : AlgebraicIndependent R x) (a : A)
+theorem algebraicIndependent_of_finite_type'
     (hinj : Injective (algebraMap R A))
-    (H : ∀ t ⊆ s, t.Finite → ∀ a ∈ s, a ∉ t → Transcendental (adjoin R t) a) :
-    AlgebraicIndependent R ((↑) : s → A) := by
+    (H : ∀ t : Set ι, t.Finite → AlgebraicIndependent R (fun i : t ↦ x i) →
+      ∀ i : ι, i ∉ t → Transcendental (adjoin R (x '' t)) (x i)) :
+    AlgebraicIndependent R x := by
   classical
-  refine algebraicIndependent_of_finite s fun t hts hfin ↦ hfin.induction_on'
-    ((algebraicIndependent_empty_iff R A).2 hinj) fun {a} {u} ha hu ha' h ↦ ?_
-  convert ((Subtype.range_coe ▸ h.option_iff a).2 <| H u (hu.trans hts) (hfin.subset hu)
-    a (hts ha) ha').comp _ (Set.subtypeInsertEquivOption ha').injective
-  ext x
+  refine algebraicIndependent_of_finite_type fun t hfin ↦ hfin.induction_on'
+    (algebraicIndependent_empty_type_iff.mpr hinj) fun {a u} ha hu ha' h ↦ ?_
+  convert ((Set.image_eq_range _ _ ▸ h.option_iff <| x a).2 <| H u (hfin.subset hu) h _ ha').comp _
+    (Set.subtypeInsertEquivOption ha').injective with x
   by_cases h : ↑x = a <;> simp [h, Set.subtypeInsertEquivOption]
 
-/-- `Type` version of `algebraicIndependent_of_finite'`. -/
-theorem algebraicIndependent_of_finite_type'
+/-- Variant of `algebraicIndependent_of_finite` using `Transcendental`. -/
+theorem algebraicIndependent_of_finite' (s : Set A)
     (hinj : Injective (algebraMap R A))
-    (H : ∀ t : Set ι, t.Finite → ∀ i : ι, i ∉ t → Transcendental (adjoin R (x '' t)) (x i)) :
-    AlgebraicIndependent R x := by
-  nontriviality R
-  haveI := hinj.nontrivial
-  have hx : Injective x := by
-    simp_rw [Transcendental] at H
-    contrapose! H
-    obtain ⟨i, j, h1, h2⟩ := not_injective_iff.1 H
-    refine ⟨{j}, by simp, i, by simp [h2], ?_⟩
-    rw [h1, Set.image_singleton]
-    exact isAlgebraic_algebraMap (⟨x j, Algebra.self_mem_adjoin_singleton R _⟩ : adjoin R {x j})
-  rw [← Set.coe_comp_rangeFactorization x]
-  refine .comp (algebraicIndependent_of_finite' _ hinj fun t ht hfin a ha ha' ↦ ?_) _
-    (Set.rightInverse_rangeSplitting hx).injective
-  change Finite t at hfin
-  have := H (x ⁻¹' t) (Finite.of_injective _ (hx.restrictPreimage t))
-    ((Equiv.ofInjective _ hx).symm ⟨_, ha⟩)
-    (by rwa [Set.mem_preimage, Equiv.apply_ofInjective_symm hx])
-  rwa [Set.image_preimage_eq_inter_range, Set.inter_eq_self_of_subset_left ht,
-    Equiv.apply_ofInjective_symm hx] at this
+    (H : ∀ t ⊆ s, t.Finite → AlgebraicIndependent R ((↑) : t → A) →
+      ∀ a ∈ s, a ∉ t → Transcendental (adjoin R t) a) :
+    AlgebraicIndependent R ((↑) : s → A) :=
+  algebraicIndependent_of_finite_type' hinj fun t hfin h i hi ↦ H _
+    (by rintro _ ⟨x, _, rfl⟩; exact x.2) (hfin.image _) h.image _ i.2
+    (mt Subtype.val_injective.mem_set_image.mp hi)
+
+namespace AlgebraicIndependent
+
+variable (hx : AlgebraicIndependent R x)
+include hx
+
+theorem adjoin_of_disjoint {s t : Set ι} (h : Disjoint s t) :
+    AlgebraicIndependent (adjoin R (x '' s)) fun i : t ↦ x i := by
```

Comment:
Maybe you can use `algebraicIndependent_of_finite'`, `aevalEquiv` and [MvPolynomial.transcendental_supported_X](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Algebraic/MvPolynomial.html#MvPolynomial.transcendental_supported_X) to prove this result. I'm not sure if the proof will be shorter or longer, but you need to import a new file.

Reply:
Changing the base ring (from `adjoin` to `supported`) is tricky, but it's apparently an art you mastered (many `comp_eq` lemmas I recall ...) I'd rather not go that route ...

Reply:
>  I'd rather not go that route

That's OK.